src/context/node.js

```diff
@@ -0,0 +1,414 @@
+   * affected nodes.
+   *
+   * See `Element.assignedSlot` API.
+   *
+   * @param {!Node} node The target node.
+   * @param {!Node} slot The slot to which the target node is assigned.
+   */
+  static assignSlot(node, slot) {
+    if (node[ASSIGNED_SLOT_PROP] == slot) {
+      return;
+    }
+    node[ASSIGNED_SLOT_PROP] = devAssert(slot);
+    forEachContained(node, (cn) => cn.discover());
+  }
+
+  /**
+   * Unassigns the direct slot previously done by the `assignSlot` call.
+   * Automatically starts the discovery phase for the affected nodes.
+   *
+   * @param {!Node} node The target node.
+   * @param {!Node} slot The slot from which the target node is assigned.
+   */
+  static unassignSlot(node, slot) {
+    if (node[ASSIGNED_SLOT_PROP] != slot) {
+      return;
+    }
+    delete node[ASSIGNED_SLOT_PROP];
+    forEachContained(node, (cn) => cn.discover());
+  }
+
+  /**
+   * Creates the context node and automatically starts the discovery process.
+   *
+   * @param {*} privateOnly
+   * @param {!Node} node
+   */
+  constructor(privateOnly, node) {
+    devAssert(privateOnly === PRIVATE_ONLY);
+
+    /** @const @private {!Node} */
+    this.node_ = node;
+
+    /** @private {?ContextNode} */
+    this.parent_ = null;
+
+    /** @private {boolean} */
+    this.parentOverriden_ = false;
+
+    /** @private {boolean} */
+    this.isRoot_ = node.nodeType == DOCUMENT_NODE;
+
+    /** @private {?ContextNode} */
+    this.root_ = this.isRoot_ ? this : null;
+
+    /** @private {?Array<!ContextNode>} */
+    this.children_ = null;
+
+    /** @const @private {function()} */
+    this.scheduleDiscover_ = oneAtATime(this.discover_.bind(this), setTimeout);
+
+    this.discover();
+  }
+
+  /**
+   * The DOM node corresponding to the context node.
+   *
+   * @return {!Node}
+   */
+  get node() {
+    return this.node_;
+  }
+
+  /**
+   * The DOM element corresponding to the context node.
+   *
+   * @return {?Element}
+   */
+  get element() {
+    // It's convenient to access Element type. If it's not an element,
+    // let it fail downstream. Otherwise, a lot of debug tools fail on trying
+    // to call JSON.stringify.
+    return this.node_.nodeType == ELEMENT_NODE
+      ? dev().assertElement(this.node_)
+      : null;
+  }
+
+  /**
+   * Whether this node is a root. The Document DOM nodes are automatically
+   * considered as roots. But other nodes can become roots as well (e.g. shadow
+   * roots) via `setIsRoot()` API.
+   *
+   * @return {boolean}
+   */
+  get isRoot() {
+    return this.isRoot_;
+  }
+
+  /**
+   * The root context node. Always available for a DOM node connected to a root
+   * node after the discovery phase.
+   *
+   * @return {?ContextNode}
+   */
+  get root() {
+    return this.root_;
+  }
+
+  /**
+   * Requests the discovery phase. Asynchronously finds the nearest parent for
+   * this node and its root. Roots and parents set directly via `setParent()`
+   * API are not discoverable.
+   */
+  discover() {
```

Comment:
I do not think we should be exposing this API publicly. The public API is just the static methods. All of the instance methods are private, and handled by us.

Reply:
This auto-discovers a number of things, but not everything, unfortunately. I think trying to discovery everything automatically will be really expensive and would involve catch-all mutation observer over the whole document and all shadow roots. So I thought we'd do limited auto-discovery and will eventually put `discover()` call into `CustomElement.connectedCallback` and `disconnectedCallback`. So, it's a restricted API, but should be public, imho.